plugins/modules/ec2_group.py

```diff
@@ -596,25 +606,39 @@ def get_target_from_rule(module, client, rule, name, group, groups, vpc_id):
     AWS accepts an ip range or a security group as target of a rule. This
     function validate the rule specification and return either a non-None
     group_id or a non-None ip range.
+
+    When using a security group as a target all 3 fields (OwnerId, GroupId, and
+    GroupName) need to exist in the target. This ensures consistency of the
+    values that will be compared to current_rules (from current_ingress and
+    current_egress) in wait_for_rule_propagation().
     """
     FOREIGN_SECURITY_GROUP_REGEX = r'^([^/]+)/?(sg-\S+)?/(\S+)'
+    owner_id = None
     group_id = None
     group_name = None
     target_group_created = False
 
     validate_rule(module, rule)
     if rule.get('group_id') and re.match(FOREIGN_SECURITY_GROUP_REGEX, rule['group_id']):
         # this is a foreign Security Group. Since you can't fetch it you must create an instance of it
+        # Matches on groups like amazon-elb/sg-5a9c116a/amazon-elb-sg, amazon-elb/amazon-elb-sg,
+        # and peer-VPC groups like 0987654321/sg-1234567890/example
         owner_id, group_id, group_name = re.match(FOREIGN_SECURITY_GROUP_REGEX, rule['group_id']).groups()
         group_instance = dict(UserId=owner_id, GroupId=group_id, GroupName=group_name)
         groups[group_id] = group_instance
         groups[group_name] = group_instance
-        # group_id/group_name are mutually exclusive - give group_id more precedence as it is more specific
         if group_id and group_name:
-            group_name = None
+            if group_name.startswith('amazon-'):
+                # amazon-elb and amazon-prefix rules don't need group_id specified,
+                group_id = None
+            else:
+                # group_id/group_name are mutually exclusive - give group_id more precedence as it is more specific
```

Comment:
Something about this feels wrong.

Are we setting ourselves up for weird edge-cases by not properly validating 'rules' as we're passed them?  The documentation states that these options are supposed to be mutually exclusive and not that one or other takes precedence.

Reply:
The terms are definitely overloaded, this section deals with the rule, and only when someone specifies what I've been thinking of as a full path or slash notated rule.group_id that matches `FOREIGN_SECURITY_GROUP_REGEX`.  This is typically (but not limited to) either a classic ELB or Peered VPC group.  And for rules, boto says: [ You must specify either the security group ID or the security group name in the request.](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.SecurityGroup.authorize_ingress)
```
  - name: Create SG from SG-ID
    ec2_group:
      name: testpeer
      description: "testcase security group"
      rules:
      - proto: tcp
        ports: 8080
        group_id: 0987654321/sg-1234567890/example
```
This Rule's `target` tuple would look like `('0987654321', 'sg-1234567890', 'example')` if we didn't set `group_name` and `owner_id` to None.

The mutual exclusivity in [the module docs](https://github.com/ansible-collections/amazon.aws/blob/59dc42e5298b89c1884fb5f4afd735c7044b9bd2/plugins/modules/ec2_group.py#L29) refers to the `name` or `group_id` of the group being created or updated - `name: testpeer` in my example above.  This code block is acting on the rule - `group_id: 0987654321/sg-1234567890/example` above.